File: python/testing/test_mht_testdrive.py
"""
MHT Test Drive: Banerjee et al. (Science 2015)
Viviano, Wuthrich, and Niehaus (2026), "A Model of Multiple Hypothesis Testing"

Requires: pandas (for reading .dta files)

To run (needs pandas):
    cd python
    pip install pandas        # plus numpy, scipy
    python testing/test_mht_testdrive.py

Data: place the Banerjee files in the repo's top-level testing/data/
(see testing/README.md; Harvard Dataverse DOI 10.7910/DVN/NHIXNT).
"""

import os
import sys
import numpy as np
import pandas as pd
from scipy.stats import norm

# Auto-detect package root (the python/ directory containing src/mhtopt/)
if os.path.exists("src/mhtopt/__init__.py"):          # run from python/
    root = os.getcwd()
elif os.path.exists("../src/mhtopt/__init__.py"):     # run from python/testing/
    root = os.path.abspath("..")
else:
    raise RuntimeError("Please run from the python/ directory or python/testing/")
sys.path.insert(0, os.path.join(root, "src"))

from mhtopt import mht_critical, mht_test, mht_est, mht_table
from mhtopt.table import print_table

# Shared validation data lives at the repo's top-level testing/data/
dta_path = os.path.join(root, "..", "testing", "data")

print()
print("=" * 70)
print("  MHT Test Drive: Banerjee et al. (Science 2015)")
print("  Viviano, Wuthrich, and Niehaus (2026)")
print("=" * 70)

# ============================================================================
# Reference table
# ============================================================================
print("\n--- Reference table ---\n")
tbl = mht_table(alpha_bar=[0.05], J_range=[1, 2, 3, 5, 6, 10],
                nm_ratios=[0.5, 1.0, 2.0])
print_table(tbl)

# ============================================================================
# VERSION A: 10 Outcomes, India
# ============================================================================
print()
print("=" * 50)
print("  VERSION A: 10 Outcomes, India")
print("  Reference: Table 3, Banerjee et al. (2015)")
print("=" * 50)

hh_outcomes = ["index_ctotal", "ind_increv", "asset_index",
               "index_foodsecurity", "ind_fin"]
mb_outcomes = ["index_time", "index_health", "index_mental",
               "index_political", "index_women"]
outcome_labels = ["Consumption", "Income", "Assets", "FoodSec", "Finance",
                  "TimeUse", "PhysHealth", "MentHealth", "Political", "WomenEmp"]

# Load data
hh_data = pd.read_stata(os.path.join(dta_path, "index_hh_vars.dta"))
hh_india = hh_data[hh_data["country"] == 4].copy()

mb_data = pd.read_stata(os.path.join(dta_path, "index_mb_vars.dta"))
mb_india = mb_data[mb_data["country"] == 4].copy()

control_hh = [c for c in hh_india.columns if c.startswith("control_")]
control_mb = [c for c in mb_india.columns if c.startswith("control_")]


def run_regression_collect_pval(data, outcome_var, control_cols):
    """Run OLS and return one-sided p-value for treatment coefficient."""
    end_var = f"{outcome_var}_end"
    bsl_var = f"{outcome_var}_bsl"
    m_bsl = f"m_{outcome_var}_bsl"
    m_country_bsl = f"m_country_{outcome_var}_bsl"

    # Ensure baseline vars exist
    for aux in [bsl_var, m_bsl, m_country_bsl]:
        if aux not in data.columns:
            data[aux] = 0.0

    rhs_cols = ["treatment", bsl_var, m_bsl, m_country_bsl] + control_cols
    rhs_cols = [c for c in rhs_cols if c in data.columns]

    y = pd.to_numeric(data[end_var], errors="coerce").values.astype(float)
    X = data[rhs_cols].apply(pd.to_numeric, errors="coerce").values.astype(float)
    X = np.column_stack([np.ones(len(y)), X])

    # Drop rows with NaN
    mask = ~(np.isnan(y) | np.any(np.isnan(X), axis=1))
    y, X = y[mask], X[mask]

    beta_hat = np.linalg.lstsq(X, y, rcond=None)[0]
    e = y - X @ beta_hat
    n_obs = len(y)
    k = X.shape[1]
    s2 = np.sum(e**2) / (n_obs - k)
    se = np.sqrt(s2 * np.diag(np.linalg.pinv(X.T @ X)))

    # treatment is column index 1 (after intercept)
    coef = beta_hat[1]
    se_val = se[1]
    t_val = coef / se_val

    # One-sided p-value (positive direction)
    if t_val >= 0:
        p_one = 1 - norm.cdf(t_val)
    else:
        p_one = 1 - (1 - norm.cdf(abs(t_val))) / 2

    return coef, se_val, t_val, p_one


# Collect results
results = []
for var in hh_outcomes:
    coef, se, t, p = run_regression_collect_pval(hh_india, var, control_hh)
    results.append((var, coef, se, t, p))
for var in mb_outcomes:
    coef, se, t, p = run_regression_collect_pval(mb_india, var, control_mb)
    results.append((var, coef, se, t, p))

p_all = np.array([r[4] for r in results])

print("\nRegression summary (India, treatment, one-sided p):")
print(f"  {'Outcome':>12}  {'Coef':>8}  {'SE':>8}  {'t':>8}  {'p':>10}")
for i, (var, coef, se, t, p) in enumerate(results):
    print(f"  {outcome_labels[i]:>12}  {coef:8.4f}  {se:8.4f}  {t:8.3f}  {p:10.6f}")

# --- mht_test: Linear model ---
print("\n--- mht_test: Linear model (cf_share=0.46), alpha_bar=0.05 ---\n")
r_lin = mht_test(p=p_all, alpha_bar=0.05)
print(f"  alpha_opt = {r_lin['alpha_opt']:.6f}")
print(f"  Rejections: opt={r_lin['reject_optimal'].sum()}, "
      f"bonf={r_lin['reject_bonferroni'].sum()}, "
      f"holm={r_lin['reject_holm'].sum()}, "
      f"bh={r_lin['reject_bh'].sum()}, "
      f"unadj={r_lin['reject_unadjusted'].sum()}")

# --- mht_test: Cobb-Douglas ---
print("\n--- mht_test: Cobb-Douglas (beta=0.13, iota=0.075) ---\n")
r_cd = mht_test(p=p_all, alpha_bar=0.05, model="cobbdouglas")
print(f"  alpha_opt = {r_cd['alpha_opt']:.6f}")
print(f"  Rejections: opt={r_cd['reject_optimal'].sum()}")

# --- Study-specific calibration ---
print("\n--- Study-specific calibration (Table 4 cost data) ---")
J = 10
for cf in [0.46, 0.75, 0.82, 0.90]:
    r = mht_critical(J=J, alpha_bar=0.05, cf_share=cf)
    print(f"  cf_share={cf:.2f}: alpha* = {r['alpha_opt']:.5f}")
print(f"  Bonferroni:     alpha* = {0.05/J:.5f}")


# ============================================================================
# VERSION B: 6 Countries as 6 Treatments
# ============================================================================
print("\n\n" + "=" * 50)
print("  VERSION B: 6 Countries as 6 Treatments")
print("  Reference: Figure 3, Table 4, Banerjee et al. (2015)")
print("=" * 50)

hh_all = hh_data.copy()
for c in range(1, 7):
    hh_all[f"treat_c{c}"] = ((hh_all["treatment"] == 1) & (hh_all["country"] == c)).astype(float)

treat_vars = [f"treat_c{c}" for c in range(1, 7)]


def run_country_mht(data, outcome_var, label, control_cols, alpha_bar=0.05):
    """Run 6-country regression and apply mht_est."""
    end_var = f"{outcome_var}_end"
    bsl_var = f"{outcome_var}_bsl"
    m_bsl = f"m_{outcome_var}_bsl"

    for aux in [bsl_var, m_bsl]:
        if aux not in data.columns:
            data[aux] = 0.0

    rhs_cols = treat_vars + [bsl_var, m_bsl] + control_cols
    rhs_cols = [c for c in rhs_cols if c in data.columns]

    y = pd.to_numeric(data[end_var], errors="coerce").values.astype(float)
    X_data = data[rhs_cols].apply(pd.to_numeric, errors="coerce").values.astype(float)
    X = np.column_stack([np.ones(len(y)), X_data])

    mask = ~(np.isnan(y) | np.any(np.isnan(X), axis=1))
    y, X = y[mask], X[mask]

    n_obs = len(y)
    k = X.shape[1]
    beta_hat = np.linalg.lstsq(X, y, rcond=None)[0]
    e = y - X @ beta_hat
    s2 = np.sum(e**2) / (n_obs - k)
    se = np.sqrt(s2 * np.diag(np.linalg.pinv(X.T @ X)))

    from scipy.stats import t as t_dist
    t_stats = beta_hat / se
    p_values = 2 * t_dist.sf(np.abs(t_stats), df=n_obs - k)

    # Build dict fit for treat_vars (indices 1..6 after intercept)
    all_names = ["const"] + rhs_cols
    fit = {
        "index": all_names,
        "params": beta_hat,
        "std_errors": se,
        "t_stats": t_stats,
        "p_values": p_values,
        "nobs": n_obs,
    }

    print(f"\n--- {label}: 6 country-specific treatments ---")
    print("    (1=Ethiopia, 2=Ghana, 3=Honduras, 4=India, 5=Pakistan, 6=Peru)")

    r = mht_est(fit, vars=treat_vars, alpha_bar=alpha_bar)
    print(f"\n  FDA default (cf_share=0.46): alpha_opt = {r['alpha_opt']:.6f}")
    for i, term in enumerate(r["terms"]):
        print(f"    {term}  coef={r['estimates'][i]:7.4f}  "
              f"p={r['p_values'][i]:.4f}  "
              f"opt={'*' if r['reject_optimal'][i] else '.'}")

    r2 = mht_est(fit, vars=treat_vars, alpha_bar=alpha_bar, cf_share=0.10)
    print(f"\n  Study-specific (cf_share=0.10): alpha_opt = {r2['alpha_opt']:.6f}")
    for i, term in enumerate(r2["terms"]):
        print(f"    {term}  coef={r2['estimates'][i]:7.4f}  "
              f"p={r2['p_values'][i]:.4f}  "
              f"opt={'*' if r2['reject_optimal'][i] else '.'}")


run_country_mht(hh_all, "index_ctotal", "Consumption", control_hh)
run_country_mht(hh_all, "ind_increv", "Income & revenues", control_hh)
run_country_mht(hh_all, "asset_index", "Assets", control_hh)
run_country_mht(hh_all, "index_foodsecurity", "Food security", control_hh)
run_country_mht(hh_all, "ind_fin", "Financial inclusion", control_hh)


# ============================================================================
# Summary
# ============================================================================
print("\n\n" + "=" * 50)
print("  SUMMARY: Version A vs Version B")
print("=" * 50)
print()
print("  Version A (10 outcomes, India):")
print("    Cost structure: adding outcomes is CHEAP (cf_share ~ 0.82)")
print("    => Strict correction needed (alpha* ~ 0.008, near Bonferroni)")
print("    => Confirms 5 core results; physical health is borderline")
print()
print("  Version B (6 countries, pooled):")
print("    Cost structure: adding a country is EXPENSIVE (cf_share ~ 0.10)")
print("    => Minimal correction needed (alpha* ~ 0.040, near unadjusted)")
print("    => Retains effects that Bonferroni would discard")
print()
print("  Same data, two framings, opposite corrections.")
print("  This is the core insight of VWN (2026).")
print("\nDone.")

File: r/testing/test_mht_testdrive.R
# ============================================================================
# MHT Test Drive: Banerjee et al. (Science 2015)
# Viviano, Wuthrich, and Niehaus (2026), "A Model of Multiple Hypothesis Testing"
#
# REFERENCE RESULTS IN BANERJEE ET AL.:
#   Table 3: Pooled treatment effects on 10 indexed outcome families, with
#            BH q-values (column 2) for all 10 hypotheses jointly.
#   Figure 3: Country-specific treatment effects at endline 2.
#   Table 4: Cost-benefit analysis with per-country cost breakdowns.
#
# VERSION A: India arm, 10 outcomes (mht_test)
#   Question: Which of the 10 outcome results survive economically-grounded MHT?
#
# VERSION B: Pooled 6-country, treatment x country (mht_est)
#   Question: Which country-specific effects survive when treating 6 programs
#             as 6 distinct treatments?
#
# REQUIREMENTS: haven (for reading .dta files)
# ============================================================================

# --- Setup ---
library(haven)

# Auto-detect package root (the r/ directory containing mhtopt/R/)
if (file.exists("mhtopt/R/mht_critical.R")) {          # run from r/
  root <- getwd()
} else if (file.exists("../mhtopt/R/mht_critical.R")) { # run from r/testing/
  root <- normalizePath("..")
} else {
  stop("Please run from the r/ directory or r/testing/")
}

# Source package functions
for (f in list.files(file.path(root, "mhtopt/R"), full.names = TRUE, pattern = "\\.R$")) {
  source(f)
}

# Shared validation data lives at the repo's top-level testing/data/
dta_path <- file.path(root, "..", "testing/data")

cat("\n")
cat(strrep("=", 70), "\n")
cat("  MHT Test Drive: Banerjee et al. (Science 2015)\n")
cat("  Viviano, Wuthrich, and Niehaus (2026)\n")
cat(strrep("=", 70), "\n")

# ============================================================================
# Reference table (cf. Table 1 in VWN 2026)
# ============================================================================
cat("\n--- Reference table ---\n\n")
tbl <- mht_table(alpha_bar = 0.05, J_range = c(1, 2, 3, 5, 6, 10),
                 nm_ratios = c(0.5, 1.0, 2.0))
print(tbl)

# ============================================================================
# VERSION A: Heterogeneous Audience -- 10 Outcomes, India
# ============================================================================
cat("\n")
cat(strrep("=", 50), "\n")
cat("  VERSION A: 10 Outcomes, India\n")
cat("  Reference: Table 3, Banerjee et al. (2015)\n")
cat(strrep("=", 50), "\n\n")

# --- Data prep: run 10 regressions, collect p-values ---
hh_outcomes <- c("index_ctotal", "ind_increv", "asset_index",
                 "index_foodsecurity", "ind_fin")
mb_outcomes <- c("index_time", "index_health", "index_mental",
                 "index_political", "index_women")
outcome_labels <- c("Consumption", "Income", "Assets", "FoodSec", "Finance",
                    "TimeUse", "PhysHealth", "MentHealth", "Political", "WomenEmp")

# HH-level outcomes
hh_data <- read_dta(file.path(dta_path, "index_hh_vars.dta"))
hh_data <- hh_data[hh_data$country == 4, ]

# MB-level outcomes
mb_data <- read_dta(file.path(dta_path, "index_mb_vars.dta"))
mb_data <- mb_data[mb_data$country == 4, ]

# Collect p-values from regressions
# (simplified: OLS with available controls, mirroring the Stata approach)
collect_pvals <- function(data, outcomes) {
  results <- list()
  control_cols <- grep("^control_", names(data), value = TRUE)

  for (var in outcomes) {
    end_var <- paste0(var, "_end")
    bsl_var <- paste0(var, "_bsl")
    m_bsl <- paste0("m_", var, "_bsl")
    m_country_bsl <- paste0("m_country_", var, "_bsl")

    # Ensure baseline vars exist (set to 0 if missing)
    for (aux in c(bsl_var, m_bsl, m_country_bsl)) {
      if (!aux %in% names(data)) data[[aux]] <- 0
    }

    # Build formula: end ~ treatment + baseline + controls
    rhs_vars <- c("treatment", bsl_var, m_bsl, m_country_bsl, control_cols)
    rhs_vars <- rhs_vars[rhs_vars %in% names(data)]
    fml <- as.formula(paste(end_var, "~", paste(rhs_vars, collapse = " + ")))

    fit <- lm(fml, data = data)
    s <- coef(summary(fit))

    coef_val <- s["treatment", "Estimate"]
    se_val <- s["treatment", "Std. Error"]
    t_val <- coef_val / se_val

    # One-sided p-value (positive direction)
    if (t_val >= 0) {
      p_one <- 1 - pnorm(t_val)
    } else {
      p_one <- 1 - (1 - pnorm(abs(t_val))) / 2
    }

    results[[var]] <- data.frame(
      outcome = var, coef = coef_val, se = se_val,
      t_stat = t_val, p_value = p_one
    )
  }
  do.call(rbind, results)
}

hh_results <- collect_pvals(hh_data, hh_outcomes)
mb_results <- collect_pvals(mb_data, mb_outcomes)
all_results <- rbind(hh_results, mb_results)
all_results$label <- outcome_labels
rownames(all_results) <- NULL

cat("Regression summary (India, treatment coefficient, one-sided p):\n")
print(all_results[, c("label", "coef", "se", "t_stat", "p_value")], row.names = FALSE)

# --- mht_test: Linear model (FDA calibration) ---
cat("\n--- mht_test: Linear model (cf_share=0.46), alpha_bar=0.05 ---\n\n")
result_linear <- mht_test(p = all_results$p_value, alpha_bar = 0.05)
print(result_linear)

# --- mht_test: Cobb-Douglas model ---
cat("\n--- mht_test: Cobb-Douglas (beta=0.13, iota=0.075) ---\n\n")
result_cd <- mht_test(p = all_results$p_value, alpha_bar = 0.05, model = "cobbdouglas")
print(result_cd)

# --- Comparison ---
cat("\n--- Comparison with Banerjee et al. Table 3 ---\n")
cat("  Paper uses BH q-values (FDR control) on pooled 6-country sample.\n")
cat("  Our analysis: India only, with cost-based optimal correction.\n\n")
cat("  Paper Table 3 q-values (endline 1, pooled):\n")
cat("    Most indices: q = 0.001  => significant under BH\n")
cat("    Physical health: q = 0.078 => NOT significant at 5% under BH\n")
cat("    Women's empowerment: q = 0.049 => borderline under BH\n\n")

# --- Study-specific calibration ---
cat("\n--- Study-specific calibration (Table 4 cost data) ---\n")
cat("  From Table 4 (India): program+overhead ~$1,268, survey ~$100-150\n")
cat("  => cf_share_outcomes ~ 0.80-0.90 (most costs fixed w.r.t. outcomes)\n\n")

J <- 10
for (cf in c(0.46, 0.75, 0.82, 0.90)) {
  r <- mht_critical(J = J, alpha_bar = 0.05, cf_share = cf)
  cat(sprintf("  cf_share=%.2f: alpha* = %.5f\n", cf, r$alpha_opt))
}
cat(sprintf("  Bonferroni:     alpha* = %.5f\n", 0.05 / J))


# ============================================================================
# VERSION B: Multiple Treatments -- 6 Countries as 6 Programs
# ============================================================================
cat("\n\n")
cat(strrep("=", 50), "\n")
cat("  VERSION B: 6 Countries as 6 Treatments\n")
cat("  Reference: Figure 3, Table 4, Banerjee et al. (2015)\n")
cat(strrep("=", 50), "\n\n")

# Load pooled data and create country-treatment indicators
hh_all <- read_dta(file.path(dta_path, "index_hh_vars.dta"))

for (c in 1:6) {
  hh_all[[paste0("treat_c", c)]] <- as.numeric(hh_all$treatment == 1 & hh_all$country == c)
}

treat_vars <- paste0("treat_c", 1:6)
control_cols <- grep("^control_", names(hh_all), value = TRUE)

# Function to run country-treatment regression and apply mht_est
run_country_mht <- function(data, outcome_var, label, alpha_bar = 0.05) {
  end_var <- paste0(outcome_var, "_end")
  bsl_var <- paste0(outcome_var, "_bsl")
  m_bsl <- paste0("m_", outcome_var, "_bsl")

  for (aux in c(bsl_var, m_bsl)) {
    if (!aux %in% names(data)) data[[aux]] <- 0
  }

  rhs_vars <- c(treat_vars, bsl_var, m_bsl, control_cols)
  rhs_vars <- rhs_vars[rhs_vars %in% names(data)]
  fml <- as.formula(paste(end_var, "~", paste(rhs_vars, collapse = " + ")))

  fit <- lm(fml, data = data)

  cat(sprintf("\n--- %s: 6 country-specific treatments ---\n", label))
  cat("    (1=Ethiopia, 2=Ghana, 3=Honduras, 4=India, 5=Pakistan, 6=Peru)\n\n")

  cat("  FDA default (cf_share=0.46):\n")
  r1 <- mht_est(fit, vars = treat_vars, alpha_bar = alpha_bar)
  print(r1)

  cat("\n  Study-specific (cf_share=0.10, countries are expensive):\n")
  r2 <- mht_est(fit, vars = treat_vars, alpha_bar = alpha_bar, cf_share = 0.10)
  print(r2)
}

run_country_mht(hh_all, "index_ctotal", "Consumption")
run_country_mht(hh_all, "ind_increv", "Income & revenues")
run_country_mht(hh_all, "asset_index", "Assets")
run_country_mht(hh_all, "index_foodsecurity", "Food security")
run_country_mht(hh_all, "ind_fin", "Financial inclusion")


# ============================================================================
# Summary
# ============================================================================
cat("\n\n")
cat(strrep("=", 50), "\n")
cat("  SUMMARY: Version A vs Version B\n")
cat(strrep("=", 50), "\n\n")
cat("  Version A (10 outcomes, India):\n")
cat("    Cost structure: adding outcomes is CHEAP (cf_share ~ 0.82)\n")
cat("    => Strict correction needed (alpha* ~ 0.008, near Bonferroni)\n")
cat("    => Confirms 5 core results; physical health is borderline\n\n")
cat("  Version B (6 countries, pooled):\n")
cat("    Cost structure: adding a country is EXPENSIVE (cf_share ~ 0.10)\n")
cat("    => Minimal correction needed (alpha* ~ 0.040, near unadjusted)\n")
cat("    => Retains effects that Bonferroni would discard\n\n")
cat("  Same data, two framings, opposite corrections.\n")
cat("  This is the core insight of VWN (2026).\n\n")
cat("Done.\n")
